core/src/main/scala/kafka/log/ProducerStateManager.scala

```diff
@@ -551,21 +556,33 @@ class ProducerStateManager(val topicPartition: TopicPartition,
    */
   def oldestSnapshotOffset: Option[Long] = oldestSnapshotFile.map(file => offsetFromFilename(file.getName))
 
+  private def isProducerRetained(producerIdEntry: ProducerIdEntry, logStartOffset: Long): Boolean = {
+    producerIdEntry.lastOffset >= logStartOffset
```

Comment:
Yes, we discussed this briefly before. I think we agreed that we would keep the `currentTxnFirstOffset` preserved (since it is difficult to find the next higher offset for a transaction), but we would fix first unstable offset to return no lower than log start offset. Does that sound correct?